code/tools/r68k/main.cpp

```diff
@@ -0,0 +1,543 @@
+#include <iostream>
+#include <filesystem>
+#include <stdint.h>
+#include <unistd.h>
+#include <thread>
+#include <atomic>
+#include <termios.h>
+#include <sys/select.h>
+#include <fcntl.h>
+#include <iomanip>
+#include <vector>
+
+#include "musashi/m68k.h"
+#include "musashi/m68kcpu.h"
+#include "AddressDecoder.h"
+
+using namespace std;
+
+#define DUART_IRQ	4
+#define DUART_VEC	0x45
+
+#define TICK_COUNT 0x408
+#define ECHO_ON    0x410
+#define PROMPT_ON  0x411
+#define LF_DISPLAY 0x412
+
+struct termios originalTermios;
+
+void init_term() {
+    struct termios newTermios;
+
+    tcgetattr(STDIN_FILENO, &originalTermios);
+
+    newTermios = originalTermios;
+    newTermios.c_lflag &= ~(ICANON | ECHO);
+    newTermios.c_iflag &= ~(ICRNL | INLCR);
+
+    tcsetattr(STDIN_FILENO, TCSANOW, &newTermios);
+
+    int flags = fcntl(STDIN_FILENO, F_GETFL, 0);
+    fcntl(STDIN_FILENO, F_SETFL, flags | O_NONBLOCK);
+}
+
+bool check_char() {
+    fd_set readfds;
+    FD_ZERO(&readfds);
+    FD_SET(STDIN_FILENO, &readfds);
+
+    struct timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 0;
+
+    return select(STDIN_FILENO + 1, &readfds, NULL, NULL, &timeout) > 0;
+}
+
+char read_char() {
+    char c = 0;
+    while (c == 0) {
+        read(STDIN_FILENO, &c, 1);
+    }
+    return c;
+}
+
+// easy68k helper functions
+
+#define BUF_LEN 78
+#define BUF_MAX BUF_LEN - 2
+static uint8_t buf[BUF_LEN];
+
+static uint8_t digit(unsigned char digit) {
+    if (digit < 10) {
+        return (char)(digit + '0');
+    } else {
+        return (char)(digit - 10 + 'A');
+    }
+}
+
+static char* print_unsigned(uint32_t num, uint8_t base) {
+    if (base < 2 || base > 36) {
+        buf[0] = 0;
+        return (char *)buf;
+    }
+
+    unsigned char bp = BUF_MAX;
+
+    if (num == 0) {
+        buf[bp--] = '0';
+    } else {
+        while (num > 0) {
+            buf[bp--] = digit(num % base);
+            num /= base;
+        }
+    } 
+
+    return ((char*)&buf[bp+1]);
+}
+
+extern "C" {
+    rosco::m68k::emu::AddressDecoder* __mem;
```

Comment:
For a program that's being built for the host, I think that it'd be better to avoid using implementation-reserved identifiers.

Reply:
Good shout, fixed 👍